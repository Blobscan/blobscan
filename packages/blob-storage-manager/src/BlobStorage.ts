import type { Environment } from "./env";
import { BlobStorageError, StorageCreationError } from "./errors";
import type { BlobStorageName } from "./types";

export interface BlobStorageConfig {
  chainId: number;
}

export abstract class BlobStorage {
  chainId: number;
  name: BlobStorageName;

  constructor(name: BlobStorageName, chainId: number) {
    this.name = name;
    this.chainId = chainId;
  }

  protected abstract _healthCheck(): Promise<void>;
  protected abstract _getBlob(uri: string): Promise<string>;
  protected abstract _storeBlob(hash: string, data: string): Promise<string>;

  async healthCheck(): Promise<"OK"> {
    try {
      await this._healthCheck();

      return "OK";
    } catch (err) {
      throw new BlobStorageError(
        this.constructor.name,
        "Storage is not reachable",
        err as Error
      );
    }
  }

  async getBlob(uri: string): Promise<string> {
    try {
      const blob = await this._getBlob(uri);

      return blob;
    } catch (err) {
      throw new BlobStorageError(
        this.constructor.name,
        `Failed to get blob with uri "${uri}"`,
        err as Error
      );
    }
  }

  async storeBlob(hash: string, data: string): Promise<string> {
    try {
      const res = await this._storeBlob(hash, data);

      return res;
    } catch (err) {
      throw new BlobStorageError(
        this.constructor.name,
        `Failed to store blob with hash "${hash}"`,
        err as Error
      );
    }
  }

  protected buildBlobFileName(hash: string): string {
    return `${this.chainId.toString()}/${hash.slice(2, 4)}/${hash.slice(
      4,
      6
    )}/${hash.slice(6, 8)}/${hash.slice(2)}.txt`;
  }

  static async create<T extends BlobStorage, C extends BlobStorageConfig>(
    this: new (config: C) => T,
    config: C
  ): Promise<T> {
    try {
      const blobStorage = new this(config);

      await blobStorage.healthCheck();

      return blobStorage;
    } catch (err) {
      const err_ = err as Error;
      throw new StorageCreationError(
        this.name,
        err_.message,
        err_.cause as Error
      );
    }
  }

  static async tryCreateFromEnv<
    C extends BlobStorageConfig,
    T extends BlobStorage
  >(
    this: {
      new (config: C): T;
      getConfigFromEnv(env: Partial<Environment>): C;
    },
    env: Partial<Environment>
  ): Promise<[T?, StorageCreationError?]> {
    let config: C;

    try {
      config = this.getConfigFromEnv(env);
    } catch (err) {
      const creationError = new StorageCreationError(
        this.name,
        "Failed to get config",
        err as Error
      );
      return [, creationError];
    }

    try {
      const blobStorage = new this(config);

      await blobStorage.healthCheck();

      return [blobStorage];
    } catch (err) {
      const err_ = err as Error;
      const creationError = new StorageCreationError(
        this.name,
        err_.message,
        err_.cause as Error
      );
      return [, creationError];
    }
  }

  protected static getConfigFromEnv(
    env: Partial<Environment>
  ): BlobStorageConfig {
    if (!env.CHAIN_ID) {
      throw new BlobStorageError(
        this.name,
        `No config variables found: no chain id provided`
      );
    }

    return {
      chainId: env.CHAIN_ID,
    };
  }
}
