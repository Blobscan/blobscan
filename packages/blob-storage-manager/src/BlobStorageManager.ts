import { api, SemanticAttributes } from "@blobscan/open-telemetry";

import type { BlobStorage } from "./BlobStorage";
import { BlobStorageManagerError } from "./errors";
import type { BlobStorageError } from "./errors";
import { tracer, updateBlobStorageMetrics } from "./instrumentation";
import type {
  Blob,
  BlobReference,
  BlobStorageName,
  StorageOf,
  StoreOptions,
} from "./types";
import { calculateBlobBytes, removeDuplicatedStorages } from "./utils";

export class BlobStorageManager {
  #blobStorages: BlobStorage[];

  constructor(blobStorages: BlobStorage[]) {
    if (!blobStorages.length) {
      throw new BlobStorageManagerError("No blob storages provided");
    }

    this.#blobStorages = removeDuplicatedStorages(blobStorages);
  }

  getStorage<T extends BlobStorageName>(name: T): StorageOf<T> | undefined {
    const blobStorage = this.#blobStorages.find(
      (storage): storage is StorageOf<T> => storage.name === name
    );

    return blobStorage as StorageOf<T> | undefined;
  }

  hasStorage<T extends BlobStorageName>(name: T): boolean {
    return !!this.getStorage(name);
  }

  async getBlob(
    ...blobReferences: BlobReference<BlobStorageName>[]
  ): Promise<{ data: string; storage: BlobStorageName } | null> {
    return tracer.startActiveSpan(
      "blob_storage_manager",
      {
        attributes: {
          [SemanticAttributes.CODE_FUNCTION]: "getBlob",
        },
      },
      async (span) => {
        const availableReferences = blobReferences.filter(({ storage }) =>
          this.hasStorage(storage)
        );

        try {
          const blob = await Promise.any(
            availableReferences.map(({ reference, storage: storageName }) =>
              tracer.startActiveSpan(
                "blob_storage_manager:storage",
                {
                  attributes: {
                    storage: storageName,
                    [SemanticAttributes.CODE_FUNCTION]: "getBlob",
                  },
                },
                async (storageSpan) => {
                  const start = performance.now();
                  const storage = this.getStorage(storageName);

                  if (!storage) {
                    throw new BlobStorageManagerError(
                      `Storage "${storageName}" is not available`
                    );
                  }

                  const blob = await storage
                    .getBlob(reference)
                    .then((data) => ({
                      data,
                      storage: storageName,
                    }));
                  const end = performance.now();

                  storageSpan.end();

                  updateBlobStorageMetrics({
                    storage: storageName,
                    blobSize: calculateBlobBytes(blob.data),
                    direction: "received",
                    duration: end - start,
                  });

                  return blob;
                }
              )
            )
          );

          span.end();

          return blob;
        } catch (e) {
          const message = "Failed to get blob from any of the storages";
          let err: BlobStorageManagerError;

          // Aggregate errors are thrown when all the promises fail
          if (e instanceof AggregateError) {
            err = new BlobStorageManagerError(
              message,
              e.errors as BlobStorageError[]
            );
          } else {
            err = new BlobStorageManagerError(message, e as Error);
          }

          span.setStatus({ code: api.SpanStatusCode.ERROR });
          span.recordException(err);

          throw err;
        } finally {
          span.end();
        }
      }
    );
  }

  async storeBlob(
    { data, versionedHash }: Blob,
    opts?: StoreOptions
  ): Promise<{
    references: BlobReference<BlobStorageName>[];
    errors: BlobStorageError[];
  }> {
    return tracer.startActiveSpan(
      "blob_storage_manager",
      {
        attributes: {
          [SemanticAttributes.CODE_FUNCTION]: "storeBlob",
        },
      },
      async (span) => {
        const selectedStorages = this.#getSelectedStorages(opts);

        const results = await Promise.allSettled(
          selectedStorages.map((storage) => {
            return tracer.startActiveSpan(
              "blob_storage_manager:storage",
              {
                attributes: {
                  storage: storage.name,
                  [SemanticAttributes.CODE_FUNCTION]: "storeBlob",
                },
              },
              async (storageSpan) => {
                const start = performance.now();
                const blobReference = await storage
                  .storeBlob(versionedHash, data)
                  .then<BlobReference<BlobStorageName>>((reference) => ({
                    reference,
                    storage: storage.name,
                  }));
                const end = performance.now();

                storageSpan.end();

                updateBlobStorageMetrics({
                  blobSize: calculateBlobBytes(data),
                  direction: "sent",
                  duration: end - start,
                  storage: storage.name,
                });

                return blobReference;
              }
            );
          })
        );
        const references = results
          .filter(
            (
              res
            ): res is PromiseFulfilledResult<BlobReference<BlobStorageName>> =>
              res.status === "fulfilled"
          )
          .map((res) => res.value);
        const storageErrors = results
          .filter((r) => r.status === "rejected")
          .map((r) => (r as PromiseRejectedResult).reason as BlobStorageError);

        if (!references.length) {
          const err = new BlobStorageManagerError(
            `Failed to upload blob with hash "${versionedHash}" to any storage`,
            storageErrors
          );

          span.setStatus({ code: api.SpanStatusCode.ERROR });
          span.recordException(err);
          throw err;
        }

        span.end();

        return {
          references,
          errors: storageErrors,
        };
      }
    );
  }

  #getSelectedStorages(
    { selectedStorages }: StoreOptions = { selectedStorages: [] }
  ): BlobStorage[] {
    const uniqueInputStorageNames = Array.from(new Set(selectedStorages));
    const selectedAvailableStorages = uniqueInputStorageNames.length
      ? this.#blobStorages.filter((s) =>
          uniqueInputStorageNames.includes(s.name)
        )
      : // Use all available storages if none are selected
        this.#blobStorages;

    if (
      uniqueInputStorageNames.length &&
      selectedAvailableStorages.length !== uniqueInputStorageNames.length
    ) {
      const selectdAvailableStorageNames = selectedAvailableStorages.map(
        (s) => s.name
      );
      const missingStorageNames = uniqueInputStorageNames.filter(
        (storageName) => !selectdAvailableStorageNames.includes(storageName)
      );

      throw new BlobStorageManagerError(
        `Some of the selected storages are not available: ${missingStorageNames.join(
          ", "
        )}"`
      );
    }

    return selectedAvailableStorages;
  }
}
