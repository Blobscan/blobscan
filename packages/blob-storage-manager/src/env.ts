import {
  z,
  booleanSchema,
  createEnv,
  presetEnvOptions,
  maskSensitiveData,
} from "@blobscan/zod";

export function parseEnv() {
  return createEnv({
    envOptions: {
      server: {
        BEE_DEBUG_ENDPOINT: z.string().url().optional(),
        BEE_ENDPOINT: z.string().url().optional(),
        CHAIN_ID: z.coerce.number().positive().default(1),
        FILE_SYSTEM_STORAGE_ENABLED: booleanSchema.default("false"),
        FILE_SYSTEM_STORAGE_PATH: z.string().optional(),
        GOOGLE_STORAGE_BUCKET_NAME: z.string().optional(),
        GOOGLE_STORAGE_PROJECT_ID: z.string().optional(),
        GOOGLE_SERVICE_KEY: z.string().optional(),
        GOOGLE_STORAGE_API_ENDPOINT: z.string().optional(),
        GOOGLE_STORAGE_ENABLED: booleanSchema.default("false"),
        POSTGRES_STORAGE_ENABLED: booleanSchema.default("true"),
        SWARM_STORAGE_ENABLED: booleanSchema.default("false"),
      },

      ...presetEnvOptions,
    },
    display(env) {
      console.log(
        `Blob storage manager configuration: chainId=${env.CHAIN_ID}, file_system=${env.FILE_SYSTEM_STORAGE_ENABLED} postgres=${env.POSTGRES_STORAGE_ENABLED}, gcs=${env.GOOGLE_STORAGE_ENABLED}, swarm=${env.SWARM_STORAGE_ENABLED}`
      );

      if (env.GOOGLE_STORAGE_ENABLED) {
        console.log(
          `GCS configuration: bucketName=${
            env.GOOGLE_STORAGE_BUCKET_NAME
          }, projectId=${maskSensitiveData(
            env.GOOGLE_STORAGE_PROJECT_ID
          )}, apiEndpoint=${env.GOOGLE_STORAGE_API_ENDPOINT}`
        );
      }

      if (env.SWARM_STORAGE_ENABLED) {
        console.log(
          `Swarm configuration: beeEndpoint=${env.BEE_ENDPOINT}, debugEndpoint=${env.BEE_DEBUG_ENDPOINT}`
        );
      }

      if (env.FILE_SYSTEM_STORAGE_ENABLED) {
        console.log(
          `File system configuration: blobDirPath=${env.FILE_SYSTEM_STORAGE_PATH}`
        );
      }
    },
  });
}
export const env = parseEnv();

export type Environment = typeof env;
