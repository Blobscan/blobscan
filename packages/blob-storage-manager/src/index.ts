export { BlobStorageManager } from "./BlobStorageManager";
export * from "./blob-storage-manager";
export { env } from "./env";
export type { Environment } from "./env";
export * from "./errors";
export { GoogleStorage, PostgresStorage, SwarmStorage } from "./storages";
export * from "./types";
