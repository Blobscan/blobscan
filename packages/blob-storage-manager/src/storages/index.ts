export * from "./FileSystemStorage";
export * from "./GoogleStorage";
export * from "./PostgresStorage";
export * from "./SwarmStorage";
