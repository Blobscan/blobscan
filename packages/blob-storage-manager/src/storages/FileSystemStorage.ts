import fs from "fs";
import path from "path";

import { BlobStorage } from "../BlobStorage";
import type { BlobStorageConfig } from "../BlobStorage";
import type { Environment } from "../env";
import { BlobStorageError } from "../errors";
import { BLOB_STORAGE_NAMES } from "../utils";

export interface FileSystemStorageConfig extends BlobStorageConfig {
  blobDirPath: string;
}

export class FileSystemStorage extends BlobStorage {
  blobDirPath: string;

  constructor({ blobDirPath, chainId }: FileSystemStorageConfig) {
    super(BLOB_STORAGE_NAMES.FILE_SYSTEM, chainId);

    this.blobDirPath = blobDirPath;

    if (!fs.existsSync(this.blobDirPath)) {
      fs.mkdirSync(this.blobDirPath);
    }
  }

  protected async _healthCheck(): Promise<void> {
    return Promise.resolve();
  }

  protected async _getBlob(versionedHash: string): Promise<string> {
    const blobFilePath = this.buildBlobFileName(versionedHash);

    try {
      const blobData = await fs.promises.readFile(blobFilePath, "utf-8");

      return blobData;
    } catch (error) {
      throw new Error(`Blob file ${versionedHash} missing: ${error}`);
    }
  }

  protected async _storeBlob(
    versionedHash: string,
    data: string
  ): Promise<string> {
    const blobfilePath = this.buildBlobFileName(versionedHash);
    const blobDirPath = blobfilePath.slice(0, blobfilePath.lastIndexOf("/"));

    if (!fs.existsSync(blobDirPath)) {
      fs.mkdirSync(blobDirPath, { recursive: true });
    }

    await fs.promises.writeFile(blobfilePath, data, { encoding: "utf-8" });

    return blobfilePath;
  }

  protected buildBlobFileName(versionedHash: string) {
    const blobFileName = super.buildBlobFileName(versionedHash);

    return path.join(this.blobDirPath, blobFileName);
  }

  static getConfigFromEnv(env: Partial<Environment>): FileSystemStorageConfig {
    const baseConfig = super.getConfigFromEnv(env);

    if (!env.FILE_SYSTEM_STORAGE_PATH) {
      throw new BlobStorageError(
        this.name,
        "No path provided. You must define variable FILE_SYSTEM_STORAGE_PATH in order to use the Filesystem storage"
      );
    }

    return {
      ...baseConfig,
      blobDirPath: env.FILE_SYSTEM_STORAGE_PATH,
    };
  }
}
