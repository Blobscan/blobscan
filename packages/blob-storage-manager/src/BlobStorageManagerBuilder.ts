import type { BlobStorage } from "./BlobStorage";
import { BlobStorageManager } from "./BlobStorageManager";
import type { BlobStorages } from "./BlobStorageManager";

export class BlobStorageManagerBuilder<
  SName extends string = never,
  T extends BlobStorages<SName> = { [k in SName]: never }
> {
  private constructor(
    private readonly blobStorages: T,
    private readonly chainId: number
  ) {}

  static create(chainId: number): BlobStorageManagerBuilder {
    return new BlobStorageManagerBuilder({}, chainId);
  }

  addStorage<K extends string, V extends BlobStorage | null>(
    name: K,
    blobStorage: V
  ): BlobStorageManagerBuilder<SName | K, T & { [k in K]: V }> {
    const nextStorage = { [name]: blobStorage } as { [k in K]: V };

    return new BlobStorageManagerBuilder(
      {
        ...this.blobStorages,
        ...nextStorage,
      },
      this.chainId
    );
  }

  build(): BlobStorageManager<SName, T> {
    return new BlobStorageManager(this.blobStorages, this.chainId);
  }
}
