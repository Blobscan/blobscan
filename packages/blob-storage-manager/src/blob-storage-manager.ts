import { logger } from "@blobscan/logger";

import type { BlobStorage } from "./BlobStorage";
import { BlobStorageManager } from "./BlobStorageManager";
import { env } from "./env";
import type { Environment } from "./env";
import type { BlobStorageName } from "./types";
import { BLOB_STORAGE_NAMES, createStorageFromEnv } from "./utils";

let blobStorageManager: BlobStorageManager | undefined;

function isBlobStorageEnabled(storageName: BlobStorageName) {
  const storageEnabledKey =
    `${storageName}_STORAGE_ENABLED` as keyof Environment;
  const storageEnabled = env[storageEnabledKey];

  return storageEnabled === true || storageEnabled === "true";
}

async function createBlobStorageManager() {
  const blobStorages = await Promise.all(
    Object.values(BLOB_STORAGE_NAMES).map(async (storageName) => {
      if (isBlobStorageEnabled(storageName)) {
        const [storage, storageError] = await createStorageFromEnv(storageName);

        if (storageError) {
          logger.warn(
            `${storageError.message}. Caused by: ${storageError.cause}`
          );
        }

        return storage;
      }
    })
  );

  const availableStorages = blobStorages.filter(
    (storage): storage is BlobStorage => !!storage
  );

  return new BlobStorageManager(availableStorages);
}

export async function getBlobStorageManager(): Promise<BlobStorageManager> {
  if (!blobStorageManager) {
    blobStorageManager = await createBlobStorageManager();
  }

  return blobStorageManager;
}
