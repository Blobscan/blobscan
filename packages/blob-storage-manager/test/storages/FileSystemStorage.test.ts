import fs from "fs";
import { beforeEach, describe, expect, it } from "vitest";

import { testValidError } from "@blobscan/test";

import { BlobStorageError, env } from "../../src";
import { FileSystemStorage } from "../../src/storages/FileSystemStorage";
import { BLOB_DATA, BLOB_HASH, FILE_URI } from "../fixtures";

class FileSystemStorageMock extends FileSystemStorage {
  constructor() {
    if (!env.FILE_SYSTEM_STORAGE_PATH) {
      throw new Error("FILE_SYSTEM_STORAGE_PATH is not set");
    }

    super({
      chainId: env.CHAIN_ID,
      blobDirPath: env.FILE_SYSTEM_STORAGE_PATH,
    });
  }

  buildBlobFileName(versionedHash: string) {
    return super.buildBlobFileName(versionedHash);
  }

  closeMock() {
    if (fs.existsSync(this.blobDirPath)) {
      fs.rmdirSync(this.blobDirPath, { recursive: true });
    }
  }
}

describe("FileSystemStorage", () => {
  let storage: FileSystemStorageMock;

  beforeEach(() => {
    storage = new FileSystemStorageMock();

    return () => {
      storage.closeMock();
    };
  });

  describe("constructor", () => {
    it("should create a new instance with all configuration options", () => {
      expect(storage).toBeDefined();
      expect(storage.chainId).toEqual(env.CHAIN_ID);
      expect(storage.blobDirPath).toEqual(
        env.FILE_SYSTEM_STORAGE_PATH
      );
    });
  });

  describe("healthCheck", () => {
    it("should resolve successfully", async () => {
      await expect(storage.healthCheck()).resolves.not.toThrow();
    });
  });

  describe("getBlob", () => {
    it("should return the contents of the blob", async () => {
      const blobFilePath = storage.buildBlobFileName(BLOB_HASH);
      const blobDirPath = blobFilePath.slice(0, blobFilePath.lastIndexOf("/"));

      fs.mkdirSync(blobDirPath, { recursive: true });
      fs.writeFileSync(blobFilePath, BLOB_DATA, { encoding: "utf-8" });

      const blob = await storage.getBlob(BLOB_HASH);

      expect(blob).toEqual(BLOB_DATA);
    });

    testValidError(
      "should throw a valid error if the blob file is missing",
      async () => {
        await storage.getBlob("missing-blob");
      },
      BlobStorageError,
      { checkCause: true }
    );
  });

  describe("storeBlob", () => {
    it("should return the correct blob file reference", async () => {
      const expectedFilePath = `${env.FILE_SYSTEM_STORAGE_PATH}/${FILE_URI}`;

      const filePath = await storage.storeBlob(BLOB_HASH, BLOB_DATA);

      expect(filePath).toEqual(expectedFilePath);
    });
  });

  describe("tryGetConfigFromEnv", () => {
    testValidError(
      "should throw a valid error if the blob dir path is not set",
      () => {
        FileSystemStorageMock.getConfigFromEnv({});
      },
      BlobStorageError
    );

    it("should return a config object if all required environment variables are set", () => {
      const config = FileSystemStorageMock.getConfigFromEnv({
        CHAIN_ID: env.CHAIN_ID,
        FILE_SYSTEM_STORAGE_PATH:
          env.FILE_SYSTEM_STORAGE_PATH,
      });

      expect(config).toEqual({
        chainId: env.CHAIN_ID,
        blobDirPath: env.FILE_SYSTEM_STORAGE_PATH,
      });
    });
  });
});
