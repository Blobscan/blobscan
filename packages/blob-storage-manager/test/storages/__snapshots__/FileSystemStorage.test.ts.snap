// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FileSystemStorage > getBlob > should throw a valid error if the blob file is missing 1`] = `"FileSystemStorageMock failed: Failed to get blob with uri \\"missing-blob\\""`;

exports[`FileSystemStorage > getBlob > should throw a valid error if the blob file is missing 2`] = `[Error: Blob file missing-blob missing: Error: ENOENT: no such file or directory, open '/tmp/blobscan-test-blob-dir/70118930558/ss/in/g-/ssing-blob.txt']`;

exports[`FileSystemStorage > tryGetConfigFromEnv > should throw a valid error if the blob dir path is not set 1`] = `"FileSystemStorageMock failed: No config variables found: no chain id provided"`;
